### PR TITLE
feat: Add hub NaPTAN code and common name fields to support cross-mode interchanges

### LINKED ISSUE #50: Backend: Add hub_naptan_code field to Station model
## Summary

Add `hub_naptan_code` and `hub_common_name` fields to the Station database model to support cross-mode interchange detection.

## Related To
- Parent Issue: #47 (Cross-mode interchange stations not recognising interchange lines)

## Problem
Stations with different TfL IDs at the same physical location (e.g., Seven Sisters has `910GSEVNSIS` for rail and `940GZZLUSVS` for tube) need to be recognized as the same interchange. TfL provides a `hubNaptanCode` field that links these stations.

## Solution
Add two new nullable columns to the `stations` table:
- `hub_naptan_code`: TfL's hub identifier (e.g., 'HUBSVS')
- `hub_common_name`: User-friendly hub name (e.g., 'Seven Sisters')

## Task List

- [ ] Add `hub_naptan_code` column to Station model (nullable, indexed)
- [ ] Add `hub_common_name` column to Station model (nullable)
- [ ] Create Alembic migration with proper upgrade/downgrade
- [ ] Test migration on test database
- [ ] Update Station schema exports (if any)
- [ ] Ensure pre-commit hooks pass
- [ ] Achieve 100% coverage on new code

### Database Schema Changes

File: `backend/app/models/tfl.py` (around line 67-122)

```python
class Station(BaseModel):
    # ... existing fields ...
    hub_naptan_code: Mapped[str | None] = mapped_column(
        String(50),
        nullable=True,
        index=True,  # For fast hub lookups during validation
        comment="TfL hub NaPTAN code for interchange stations"
    )
    hub_common_name: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
        comment="Common name for the hub (e.g., 'Seven Sisters')"
    )
```

### Alembic Migration

Create: `backend/alembic/versions/{timestamp}_add_hub_naptan_code.py`

```python
"""Add hub_naptan_code to stations

Revision ID: {generated}
Revises: {previous_revision}
Create Date: {timestamp}
"""
from alembic import op
import sqlalchemy as sa

revision = '{generated}'
down_revision = '{previous_revision}'
branch_labels = None
depends_on = None

def upgrade():
    op.add_column('stations', sa.Column(
        'hub_naptan_code', 
        sa.String(50), 
        nullable=True, 
        comment='TfL hub NaPTAN code'
    ))
    op.add_column('stations', sa.Column(
        'hub_common_name', 
        sa.String(255), 
        nullable=True, 
        comment='Hub common name'
    ))
    op.create_index(
        'ix_stations_hub_naptan_code', 
        'stations', 
        ['hub_naptan_code']
    )

def downgrade():
    op.drop_index('ix_stations_hub_naptan_code', 'stations')
    op.drop_column('stations', 'hub_common_name')
    op.drop_column('stations', 'hub_naptan_code')
```

## Testing Requirements

Test coverage must be 100% on new code.

Test cases:
```python
def test_migration_upgrade_adds_columns():
    """Test migration adds hub columns and index."""
    # Run upgrade
    # Verify columns exist
    # Verify index exists
    
def test_migration_downgrade_removes_columns():
    """Test migration removes hub columns and index."""
    # Run downgrade
    # Verify columns removed
    # Verify index removed
    
def test_migration_preserves_existing_data():
    """Test existing station data preserved during migration."""
    # Insert test data
    # Run migration
    # Verify data intact
    
def test_station_model_hub_fields_nullable():
    """Test hub fields accept None values."""
    # Create station without hub data
    # Verify no errors
```

## Files to Modify

- `backend/app/models/tfl.py` (Station model)
- `backend/alembic/versions/{timestamp}_add_hub_naptan_code.py` (new)
- `backend/tests/test_migrations.py` (add tests if not exists)

## Validation Steps

After implementation:
1. Run migration: `cd backend && uv run alembic upgrade head`
2. Verify columns: Check database schema
3. Verify index: Check database indexes
4. Run tests: `uv run pytest -xvs`
5. Test downgrade: `uv run alembic downgrade -1`
6. Test re-upgrade: `uv run alembic upgrade head`
7. Run pre-commit: `uv run pre-commit run --all-files`

## Dependencies

None (foundation for other issues)

## Blocks

- Issue tracking #47-B (Backend: Fetch hub data)
- Issue tracking #47-C (Backend: Route validation)

## Coverage Target

- **New Code:** 100%
- **Overall:** Maintain existing coverage (currently ~95%)

### PR DESCRIPTION
**Pull request by mnbf9rca**
closes #50

## Summary by Sourcery

Implement support for cross-mode interchanges by adding hub NaPTAN code and common name fields to stations, updating fetch logic to extract and persist this hub information, refactoring station-fetching into reusable helpers, and extending database schema, API models, documentation, and test coverage accordingly.

New Features:
- Add nullable hub_naptan_code and hub_common_name columns to the stations table (indexed) via an Alembic migration
- Include hub_naptan_code and hub_common_name in the Station ORM model and API response schema
- Extract and persist hub fields during station fetch to link related stations for cross-mode interchanges

Enhancements:
- Refactor fetch_stations into helper methods (_extract_hub_fields, _update_existing_station, _create_new_station, _fetch_stations_from_api)
- Simplify fetch_stations to delegate API fetch and database updates to the new helpers

Build:
- Add migration to add hub fields and create index on hub_naptan_code

Documentation:
- Update ADR to document hub NaPTAN code design and cross-mode interchange support

Tests:
- Add unit tests for hub field extraction, station creation/update with hub fields, and error handling in fetch methods
- Add model tests for persisting nullable hub fields
- Add integration and coverage tests for station-fetching and exception paths

### CODE REVIEW

**Comment by mnbf9rca** [comment]
## Implementation Plan for Review Comments

Addressing review comment #2508119019 from PR review.

### Task List
- [x] Implement fix: add else clause to clear hub_common_name when hub_code is None
- [x] Add test case for hub_code → None transition (clearing stale hub data)
- [x] Run tests to verify 100% coverage is maintained
- [x] Reply to and resolve review comment #2508119019

### Analysis
**Comment #2508119019**: Add else clause to clear hub_common_name when hub_code is None

**Decision**: ✅ ACCEPT
- Valid bug fix preventing stale data
- Improves consistency with new station creation path (line 621)
- Aligns with ADR requirement for proper NULL handling
- Requires additional test coverage for hub_code → None scenario

**Rationale**: Without this fix, a station that loses its hub status would retain old hub_common_name, violating data consistency (hub_common_name should only exist when hub_naptan_code exists).

Starting implementation...

File: backend/app/services/tfl_service.py [hunk 1]
```diff
@@ -600,15 +600,25 @@ async def fetch_stations(self, line_tfl_id: str | None = None, use_cache: bool =
                         if line_tfl_id not in station.lines:
                             station.lines = [*station.lines, line_tfl_id]
                         station.last_updated = datetime.now(UTC)
+                        # Update hub information if present (Place doesn't have hubNaptanCode, only StopPoint does)
+                        hub_code = getattr(stop_point, "hubNaptanCode", None)
+                        station.hub_naptan_code = hub_code
+                        if hub_code:
+                            # Use commonName as hub name when hub code exists
+                            station.hub_common_name = stop_point.commonName
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py]: ✅ Fixed. Added else clause to set hub_common_name=None when no hub code exists. This ensures consistency with new station creation and prevents stale data. All tests passing (582/582, 98% coverage).

**Comment by mnbf9rca** [comment]
## Review Comments Addressed

I've addressed all review comments as follows:

### ✅ Accepted & Implemented

1. **Missing Test Coverage** (Sourcery): Added 7 new tests
   - `test_extract_hub_fields_with_empty_string` - Edge case for empty string hub codes
   - `test_extract_hub_fields_with_whitespace` - Edge case for whitespace hub codes
   - `test_update_existing_station_updates_hub_fields` - Hub field updates (old → new)
   - `test_update_existing_station_hub_name_change_only` - Hub name changes independently
   - `test_fetch_stations_with_hub_fields` - Integration test for hub field population
   - `test_fetch_stations_without_hub_fields` - Integration test for stations without hubs
   - `test_fetch_stations_updates_changed_hub_fields` - Integration test for hub updates (old → new)
   - `test_fetch_stations_clears_hub_fields_when_removed` - Integration test for hub removal

   **Test Results**: All 141 tests passing ✅ | Coverage: 99.54% on tfl_service.py (exceeds 95% target) ✅

### ⚠️ Deferred (YAGNI Principle)

2. **N+1 Query Optimization** (Comment #2508149259)
   - **Decision**: Defer with inline code comment
   - **Rationale**: TfL lines have 20-60 stations, operation is infrequent (cached ~daily), no performance issue demonstrated
   - **Action**: Added code comment at line 646-651 explaining this is acceptable for current scale
   - **Documentation**: Added to `implementation_plan.md` "Deferred Code Quality Improvements" section

3. **Bulk Refresh Optimization** (Comment #2508149261)
   - **Decision**: Defer (current approach is clear and safe)
   - **Rationale**: Marginal benefit (~20-60ms savings), SQLAlchemy refresh is fast, explicit is better for hobby project
   - **Documentation**: Added to `implementation_plan.md` "Deferred Code Quality Improvements" section

### ❌ Rejected

4. **Use `station.lines.append()`** (Sourcery overall comment)
   - **Decision**: Reject - Current implementation is correct for JSON columns
   - **Rationale**: 
     - `station.lines` is a JSON column (`Mapped[list[str]]`)
     - Direct mutations like `.append()` don't trigger SQLAlchemy change tracking for JSON columns
     - Current approach `station.lines = [*station.lines, line_tfl_id]` creates new list → SQLAlchemy detects change ✅
     - To use `.append()` safely would require `MutableList` or `flag_modified()`, neither used in this project
   - **Evidence**: Project doesn't use `MutableList` or `flag_modified` (grep confirms)

### Summary of Changes
- ✅ **Code**: Added inline comment explaining N+1 query pattern
- ✅ **Tests**: Added 7 comprehensive tests (4 unit + 3 integration)
- ✅ **Coverage**: Maintained 99.54% on tfl_service.py
- ✅ **Documentation**: Updated `implementation_plan.md` with deferred optimizations

All tests passing, coverage excellent, ready for re-review\! 🎉

File: backend/app/services/tfl_service.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py]: ✅ **Already Fixed**

This issue was resolved prior to this review. The `else` clause was added to set `hub_common_name = None` when `hub_code` is None (see Comment #2508149954).

**Current implementation** (line 606-608 after refactoring):
```python
if hub_code:
    station.hub_common_name = stop_point.commonName
else:
    station.hub_common_name = None  # Clear stale data
```

**Test Coverage**: 
- `test_update_existing_station_clears_hub_fields` (line 814) validates this behavior
- `test_fetch_stations_clears_hub_fields_when_removed` (integration test) validates end-to-end

All tests passing ✅

File: backend/app/services/tfl_service.py [hunk 2]
```diff
@@ -544,6 +544,131 @@ async def fetch_stop_types(self, use_cache: bool = True) -> list[StopType]:
...
+        # Type narrowing: _handle_api_error raises if response is ApiError, so it's safe here
+        ttl = self._extract_cache_ttl(response) or DEFAULT_STATIONS_CACHE_TTL  # type: ignore[arg-type]
+        # response.content is a PlaceArray (RootModel), access via .root
+        stop_points = response.content.root  # type: ignore[union-attr]
+
+        stations = []
+        for stop_point in stop_points:
+            # Check if station exists in DB
+            result = await self.db.execute(select(Station).where(Station.tfl_id == stop_point.id))
+            station = result.scalar_one_or_none()
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py:649]: ⚠️ **Deferred per YAGNI Principle**

This N+1 query pattern is acceptable for the current scale of this hobby project.

**Rationale for deferring**:
- TfL lines typically have 20-60 stations (small N)
- N+1 with 60 stations = 61 queries, executes in milliseconds
- Operation is infrequent (station data is cached, runs ~once per day per line)
- Database queries are fast (local DB, indexed `tfl_id` column)
- No demonstrated performance problem

**Actions taken**:
- ✅ Added inline code comment (lines 646-651) explaining this is acceptable
- ✅ Documented in `implementation_plan.md` "Deferred Code Quality Improvements" section

**When to revisit**:
- If monitoring shows slow station fetch operations (>500ms)
- If the project scales to support multiple transit systems with hundreds of stations per line
- If station data refresh frequency increases significantly

Following KISS and YAGNI principles - optimize when there's an actual bottleneck, not speculatively.

File: backend/app/services/tfl_service.py [hunk 3]
```diff
@@ -544,6 +544,131 @@ async def fetch_stop_types(self, use_cache: bool = True) -> list[StopType]:
...
+                station = self._create_new_station(stop_point, line_tfl_id, hub_code, hub_name)
+                self.db.add(station)
+
+            stations.append(station)
+
+        await self.db.commit()
+
+        # Refresh to get database IDs
+        for station in stations:
+            await self.db.refresh(station)
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py:668]: ⚠️ **Deferred - Current Approach is Clear and Safe**

The current implementation is appropriate for this hobby project.

**Rationale for deferring**:
- **Marginal benefit**: SQLAlchemy refresh is fast (in-memory cache hit for most cases)
- **Correctness concern**: Existing stations ARE modified (lines array, last_updated, hub fields), so they may benefit from refresh to ensure ORM state matches DB state
- **Complexity increase**: Tracking new vs. updated stations adds conditional logic
- **Small scale**: 20-60 stations × refresh time (~1ms) = 20-60ms total, negligible
- **YAGNI principle**: No demonstrated performance problem

**Counter-argument**: SQLAlchemy docs suggest refresh is mainly for DB-generated values (IDs, timestamps). However, the current implementation is explicit and guarantees consistency.

**Actions taken**:
- ✅ Documented in `implementation_plan.md` "Deferred Code Quality Improvements" section

**When to revisit**: Only if profiling shows refresh as a bottleneck (unlikely).

Current implementation prioritizes clarity and safety over micro-optimization.

File: backend/tests/test_tfl_service.py [hunk 4]
```diff
@@ -4827,3 +4928,252 @@ async def test_get_station_routes_database_error(
...
+async def test_fetch_stations_update_existing_station(
+    tfl_service: TfLService,
+    db_session: AsyncSession,
+) -> None:
+    """Test fetch_stations updates existing station with new line (covers line 656)."""
+    with freeze_time("2025-01-01 12:00:00"):
+        # Create existing station with one line
+        existing_station = Station(
+            tfl_id="940GZZLUVIC",
+            name="Victoria",
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/tests/test_tfl_service.py:4945]: ✅ **Test Added**

Added comprehensive test coverage for hub field updates (old → new scenario).

**New Tests Added**:

1. **Unit Test** - `test_update_existing_station_updates_hub_fields` (line 939)
   - Tests `_update_existing_station()` with old hub → new hub
   - Verifies hub fields are overwritten, not just added

2. **Integration Test** - `test_fetch_stations_updates_changed_hub_fields` (line 5136)
   - Tests full flow: existing station with `OLDHUB` → API returns `HUBKGX`
   - Verifies hub fields UPDATED end-to-end

3. **Additional Coverage** - `test_update_existing_station_hub_name_change_only` (line 968)
   - Tests scenario where hub code stays same but name changes
   - Validates independent hub name updates

**Test Results**: All 141 tests passing ✅ | Coverage: 99.54% on tfl_service.py ✅

The implementation correctly handles hub field updates via simple assignment (`station.hub_naptan_code = hub_code`), which overwrites old values as expected.

**Comment by mnbf9rca** [comment]
## Response to Sourcery Overall Comments

### 1. ✅ Integration Test for JSON Response - **Implemented**
Added integration tests to verify hub fields are returned in fetch_stations:
- `test_fetch_stations_with_hub_fields` - Verifies hub fields populated from API
- `test_fetch_stations_without_hub_fields` - Verifies stations without hubs
- `test_fetch_stations_updates_changed_hub_fields` - Verifies hub updates
- `test_fetch_stations_clears_hub_fields_when_removed` - Verifies hub removal

### 2. ✅ Edge-Case Test for Empty/Whitespace Hub Code - **Implemented**
Added unit tests:
- `test_extract_hub_fields_with_empty_string` - Validates empty string returns None
- `test_extract_hub_fields_with_whitespace` - Documents whitespace behavior (truthy per Python rules)

### 3. ❌ Use `station.lines.append()` - **Rejected**

**This suggestion is incorrect for JSON columns in SQLAlchemy.**

**Current implementation** (line 580):
```python
if line_tfl_id not in station.lines:
    station.lines = [*station.lines, line_tfl_id]  # ✅ Correct for JSON columns
```

**Why `.append()` doesn't work**:
- `station.lines` is a JSON column (`Mapped[list[str]]` with `JSON` type)
- Direct in-place mutations like `.append()` **do not trigger SQLAlchemy change tracking** for JSON columns
- Changes wouldn't be persisted to the database
- The current approach creates a new list and reassigns it, which SQLAlchemy **does** detect as a change

**To use `.append()` safely would require**:
```python
from sqlalchemy.ext.mutable import MutableList
# OR
from sqlalchemy.orm import attributes
station.lines.append(line_tfl_id)
attributes.flag_modified(station, "lines")
```

**Evidence**: This project doesn't use `MutableList` or `flag_modified`:
```bash
$ grep -r "MutableList\|flag_modified" backend/
# No results
```

**Conclusion**: Current implementation is correct. Changing to `.append()` would introduce a **bug** where line updates aren't persisted. 

**Reference**: SQLAlchemy docs on [Mutation Tracking for Mutable Types](https://docs.sqlalchemy.org/en/20/core/custom_types.html#mutation-tracking)
